src/libraries/Common/src/Interop/BSD/System.Native/Interop.Sysctl.cs

```diff
@@ -64,21 +62,36 @@ private static unsafe void Sysctl(int* name, int name_len, ref byte* value, ref
                 }
                 else
                 {
-                    bytesLength = (int)bytesLength * 2;
+                    bytesLength *= 2;
                 }
-                value = (byte*)Marshal.AllocHGlobal(bytesLength);
+                value = (byte*)NativeMemory.Alloc(bytesLength);
                 ret = Sysctl(name, name_len, value, &bytesLength);
             }
             if (ret != 0)
             {
+                nint errno = Marshal.GetLastPInvokeError();
                 if (autoSize)
                 {
-                    Marshal.FreeHGlobal((IntPtr)value);
+                    NativeMemory.Free(value);
                 }
-                throw new InvalidOperationException(SR.Format(SR.InvalidSysctl, *name, Marshal.GetLastPInvokeError()));
+                ThrowInvalidSysctlException(name, name_len, errno);
             }
 
-            len = (int)bytesLength;
+            len = (uint)bytesLength;
+        }
+
+        private static unsafe InvalidOperationException ThrowInvalidSysctlException(int* name, uint name_len, nint errno)
+        {
+            int mib0 = name[0];
```

Comment:
Would it be better to format the full `name` to avoid extra or missing entries? You can use `string.Join(",", name.ToArray())` to do that. (This assumes that you pass in ReadOnlySpan<int> instead of pointers.)